File: python/crypto-py/pythonn/saes.cpp
#include <bits/stdc++.h>
using namespace std;

// --- Step 1: Define S-Box and Inverse S-Box ---
map<int, int> S_BOX = {
    {0x0, 0x9}, {0x1, 0x4}, {0x2, 0xA}, {0x3, 0xB},
    {0x4, 0xD}, {0x5, 0x1}, {0x6, 0x8}, {0x7, 0x5},
    {0x8, 0x6}, {0x9, 0x2}, {0xA, 0x0}, {0xB, 0x3},
    {0xC, 0xC}, {0xD, 0xE}, {0xE, 0xF}, {0xF, 0x7}
};

map<int, int> INV_S_BOX;

// --- Step 2: Helper functions ---
int sub_nibbles(int byte) {
    return (S_BOX[(byte >> 4) & 0xF] << 4) | S_BOX[byte & 0xF];
}

int inv_sub_nibbles(int byte) {
    return (INV_S_BOX[(byte >> 4) & 0xF] << 4) | INV_S_BOX[byte & 0xF];
}

int shift_rows(int word) {
    return ((word & 0xF0) >> 4) | ((word & 0x0F) << 4);
}

int gf_mult(int a, int b) {
    int p = 0;
    for (int i = 0; i < 4; i++) {
        if (b & 1) p ^= a;
        int carry = a & 0x8;
        a = (a << 1) & 0xF;
        if (carry) a ^= 0x3;
        b >>= 1;
    }
    return p & 0xF;
}

int mix_columns(int word) {
    int a0 = (word >> 12) & 0xF;
    int a1 = (word >> 8) & 0xF;
    int a2 = (word >> 4) & 0xF;
    int a3 = word & 0xF;

    int b0 = a0 ^ gf_mult(4, a2);
    int b1 = a1 ^ gf_mult(4, a3);
    int b2 = a2 ^ gf_mult(4, a0);
    int b3 = a3 ^ gf_mult(4, a1);

    return (b0 << 12) | (b1 << 8) | (b2 << 4) | b3;
}

// --- Step 3: Key Expansion (16-bit -> 3 round keys) ---
vector<int> key_expansion(int key) {
    vector<int> w(6);
    w[0] = (key >> 8) & 0xFF;
    w[1] = key & 0xFF;

    int RCON1 = 0x80, RCON2 = 0x30;

    w[2] = w[0] ^ sub_nibbles(shift_rows(w[1])) ^ RCON1;
    w[3] = w[2] ^ w[1];
    w[4] = w[2] ^ sub_nibbles(shift_rows(w[3])) ^ RCON2;
    w[5] = w[4] ^ w[3];

    int k0 = (w[0] << 8) | w[1];
    int k1 = (w[2] << 8) | w[3];
    int k2 = (w[4] << 8) | w[5];

    return {k0, k1, k2};
}

// --- Step 4: AddRoundKey ---
int add_round_key(int state, int key) {
    return state ^ key;
}

// --- Step 5: Encryption ---
int encrypt(int plaintext, int key) {
    vector<int> keys = key_expansion(key);
    int k0 = keys[0], k1 = keys[1], k2 = keys[2];

    int state = add_round_key(plaintext, k0);

    state = (sub_nibbles((state >> 8) & 0xFF) << 8) | sub_nibbles(state & 0xFF);
    state = shift_rows(state);
    state = mix_columns(state);
    state = add_round_key(state, k1);

    state = (sub_nibbles((state >> 8) & 0xFF) << 8) | sub_nibbles(state & 0xFF);
    state = shift_rows(state);
    state = add_round_key(state, k2);

    return state;
}

// --- Step 6: Decryption ---
int decrypt(int ciphertext, int key) {
    vector<int> keys = key_expansion(key);
    int k0 = keys[0], k1 = keys[1], k2 = keys[2];

    int state = add_round_key(ciphertext, k2);
    state = shift_rows(state);
    state = (inv_sub_nibbles((state >> 8) & 0xFF) << 8) | inv_sub_nibbles(state & 0xFF);

    state = add_round_key(state, k1);
    state = mix_columns(state); // same in S-AES
    state = shift_rows(state);
    state = (inv_sub_nibbles((state >> 8) & 0xFF) << 8) | inv_sub_nibbles(state & 0xFF);

    state = add_round_key(state, k0);
    return state;
}

// --- Main Function ---
int main() {
    // build inverse S-BOX
    for (auto &p : S_BOX) {
        INV_S_BOX[p.second] = p.first;
    }

    int plaintext = 0b1101011100101000; // Example 16-bit plaintext
    int key = 0b0100101011110101;       // Example 16-bit key

    cout << "Plaintext : " << bitset<16>(plaintext) << endl;
    int ciphertext = encrypt(plaintext, key);
    cout << "Ciphertext: " << bitset<16>(ciphertext) << endl;
    int decrypted = decrypt(ciphertext, key);
    cout << "Decrypted : " << bitset<16>(decrypted) << endl;

    return 0;
}

This paste expires in <1 hour. Public IP access. Share whatever you see with others in seconds with Context.Terms of ServiceReport this

File: python/crypto-py/pythonn/saes.py

S_BOX = {
    0x0: 0x9, 0x1: 0x4, 0x2: 0xA, 0x3: 0xB,
    0x4: 0xD, 0x5: 0x1, 0x6: 0x8, 0x7: 0x5,
    0x8: 0x6, 0x9: 0x2, 0xA: 0x0, 0xB: 0x3,
    0xC: 0xC, 0xD: 0xE, 0xE: 0xF, 0xF: 0x7
}
INV_S_BOX = {v: k for k, v in S_BOX.items()}


def gf_mult(a, b):
    a &= 0xF; b &= 0xF
    p = 0
    for _ in range(4):
        if b & 1:
            p ^= a
        carry = a & 0x8
        a = (a << 1) & 0xF
        if carry:
            a ^= 0x3 
        b >>= 1
    return p & 0xF

def sub_nibbles16(state):
    n0 = (state >> 12) & 0xF
    n1 = (state >> 8)  & 0xF
    n2 = (state >> 4)  & 0xF
    n3 =  state        & 0xF
    n0 = S_BOX[n0]; n1 = S_BOX[n1]; n2 = S_BOX[n2]; n3 = S_BOX[n3]
    return (n0 << 12) | (n1 << 8) | (n2 << 4) | n3

def inv_sub_nibbles16(state):
    n0 = (state >> 12) & 0xF
    n1 = (state >> 8)  & 0xF
    n2 = (state >> 4)  & 0xF
    n3 =  state        & 0xF
    n0 = INV_S_BOX[n0]; n1 = INV_S_BOX[n1]; n2 = INV_S_BOX[n2]; n3 = INV_S_BOX[n3]
    return (n0 << 12) | (n1 << 8) | (n2 << 4) | n3

def shift_rows(state):

    n0 = (state >> 12) & 0xF
    n1 = (state >> 8)  & 0xF
    n2 = (state >> 4)  & 0xF
    n3 =  state        & 0xF
    
    return (n0 << 12) | (n3 << 8) | (n2 << 4) | (n1)

def inv_shift_rows(state):

    return shift_rows(state)

def mix_columns(state):

    n0 = (state >> 12) & 0xF
    n1 = (state >> 8)  & 0xF
    n2 = (state >> 4)  & 0xF
    n3 =  state        & 0xF
    b0 = n0 ^ gf_mult(4, n2)
    b1 = n1 ^ gf_mult(4, n3)
    b2 = n2 ^ gf_mult(4, n0)
    b3 = n3 ^ gf_mult(4, n1)
    return (b0 << 12) | (b1 << 8) | (b2 << 4) | b3

def inv_mix_columns(state):
 
    n0 = (state >> 12) & 0xF
    n1 = (state >> 8)  & 0xF
    n2 = (state >> 4)  & 0xF
    n3 =  state        & 0xF
    b0 = gf_mult(9, n0) ^ gf_mult(2, n2)
    b1 = gf_mult(9, n1) ^ gf_mult(2, n3)
    b2 = gf_mult(2, n0) ^ gf_mult(9, n2)
    b3 = gf_mult(2, n1) ^ gf_mult(9, n3)
    return (b0 << 12) | (b1 << 8) | (b2 << 4) | b3

def sub_nibbles8(byte):
    return (S_BOX[(byte >> 4) & 0xF] << 4) | S_BOX[byte & 0xF]

def rot_nibbles8(byte):
    return ((byte & 0x0F) << 4) | ((byte & 0xF0) >> 4)

def key_expansion(key16):
    w0 = (key16 >> 8) & 0xFF
    w1 = key16 & 0xFF
    RCON1, RCON2 = 0x80, 0x30
    w2 = w0 ^ sub_nibbles8(rot_nibbles8(w1)) ^ RCON1
    w3 = w2 ^ w1
    w4 = w2 ^ sub_nibbles8(rot_nibbles8(w3)) ^ RCON2
    w5 = w4 ^ w3
    k0 = (w0 << 8) | w1
    k1 = (w2 << 8) | w3
    k2 = (w4 << 8) | w5
    return k0, k1, k2

def add_round_key(state, key):
    return state ^ key


def encrypt_saes(plaintext, key16):
    k0, k1, k2 = key_expansion(key16)
    print("Expanded key: ",k0,k1,k2)
    s = add_round_key(plaintext, k0)
    print("Block after added round key:",s)
    s = sub_nibbles16(s)
    print("Round 1 -> Sub bytes:",s)
    s = shift_rows(s)
    print("Round 1 -> Shift rows:",s)
    s = mix_columns(s)
    print("Round 1 -> Mix columns:",s)
    s = add_round_key(s, k1)
    print("Round 1 -> Round key k1:",s)
    s = sub_nibbles16(s)
    s = shift_rows(s)
    s = add_round_key(s, k2)
    return s & 0xFFFF


def decrypt_saes(ciphertext, key16):
    k0, k1, k2 = key_expansion(key16)
    s = add_round_key(ciphertext, k2)
    s = inv_shift_rows(s)
    s = inv_sub_nibbles16(s)
    s = add_round_key(s, k1)
    s = inv_mix_columns(s)
    s = inv_shift_rows(s)
    s = inv_sub_nibbles16(s)
    s = add_round_key(s, k0)
    return s & 0xFFFF


if __name__ == "__main__":
    P = int("1101011100101000", 2) 
    K = int("0100101011110101", 2)  
    C = encrypt_saes(P, K)
    D = decrypt_saes(C, K)
    print("Plaintext : ", format(P, "016b"))
    print("Ciphertext:", format(C, "016b"))
    print("Decrypted : ", format(D, "016b"))


#cipher=0b0010010011101100